LAYOUTS-629 Implement editing blocks in the preview

bcafuk opened the pull request.

Commit:
Add missing closing parenthesis

Commit:
Run Prettier

Commit:
LAYOUTS-629 Implement editing blocks in the preview

MarkoZabcic commented on components.js:
```diff
@@ -1,3 +1,4 @@
 export {default as Avatar} from './components/avatar/element.js';
 export {default as Badge} from './components/badge/element.js';
+export {default as Block} from './components/block/element.js';
 export {default as Dropdown} from './components/dropdown/element.js';
```
Let's entirely remove Avatar, Badge and Dropdown components as they where only examples. You will always have them in Git.

And all of the files

bcafuk replied:
Fixed in 910bf1a4bc8b0101a662ba2a0ed221a9c6503e64

MarkoZabcic commented on components/block/spec.js:
```diff
@@ -11,7 +11,9 @@ describe('ngl-block', () => {
-    assert.shadowDom.equal(element, `
+    assert.shadowDom.equal(
+      element,
+      `
```
Fix indent

bcafuk replied:
Fixed in 07f98d8290fcd308b5a249a4ec325edfe5c3028a

MarkoZabcic replied:
@bcafuk also have consistency in opening and closing quotes.

bcafuk replied:
Fixed inconsistent quotes in 560d350cbacc45f5a0b30b9289c98c1b67270490

MarkoZabcic commented on components/block/element.js:
```diff
@@ -9,56 +9,93 @@ export default class Block extends LitElement {
 
   static properties = {
-    blockId: { type: Number},
+    blockId: {type: String, attribute: 'data-ngl-block-id'},
```
Why are you using different attribute name here?

bcafuk replied:
My idea was to use dataset attributes for consistency with existing custom attributes in Layouts. Should I just use the default attribute name of `blockId`?

emodric replied:
I think that the attribute name is fine. Every custom layouts data attribute has `data-ngl-`, so for consistency sake, `data-ngl-block-id` is perfectly fine.

MarkoZabcic replied:
Previously you had to use data attributes but now you don't as you have custom tag.

With native web components you have few advantages which where impossible before:

1. You have custom tag named `<ngl-block>`
2. You have custom attribute `blockId` which also has it's property
3. See code below

```js
const block = document.querySelector('ngl-block');
console.log(block.blockId) // get
block.blockId = 33; // set block Id
block.setAttribute('blockId', 10); // or set attribute
block.getAttribute('blockId'); // or get attribute
```

In this case I would rather avoid setting blockId as data attribute and use dataset mechanism to read it as that is more confusing in this case.

bcafuk replied:
@MarkoZabcic I can see the reasoning behind that. I had picked the name purely for consistency's sake. However, the attribute isn't meant to be used through the dataset at all, so it makes sense not to prefix it with `data-`.

I have renamed the attribute in ddfec7b6db6863d2f111e70692f5000feb38a64f.

emodric replied:
:+1:

MarkoZabcic commented on components/block/element.js:
```diff
@@ -9,56 +9,93 @@ export default class Block extends LitElement {
-    console.log(e.target.innerText)
-    this.remove();
+  getBlock() {
+    return window.parent.Core.g.layout.blocks.findWhere({id: this.blockId});
```
Because we are inside of a web component called block it would be better to have different namings for getBlock.

Also as we are using es6 getters are much better suited here.


```js
get layout() {
  return window.parent.Core.g.layout;
}

get model() {
  return this.cached_model ||= this.layout.blocks.findWhere({id: this.blockId});
}
```

bcafuk replied:
Fixed in 506394a14afbb78b493e59d4dc5f6fc4a04890d8

MarkoZabcic commented on components/block/element.js:
```diff
@@ -9,56 +9,93 @@ export default class Block extends LitElement {
-    const prevElement = this.previousElementSibling;
-    isBlock(prevElement) && this.parentNode.insertBefore(this, prevElement)
+  renderMenu() {
+    const block = this.getBlock();
```
Now you can just reference `block` by `this.block` without a need for this line

bcafuk replied:
Fixed in 506394a14afbb78b493e59d4dc5f6fc4a04890d8

MarkoZabcic commented on components/block/element.js:
```diff
@@ -9,56 +9,93 @@ export default class Block extends LitElement {
-    isBlock(prevElement) && this.parentNode.insertBefore(this, prevElement)
+  renderMenu() {
+    const block = this.getBlock();
+    const isInLinkedZone = block.zone().is_linked();
```
extract to a getter

```js
get isInLinkedZone() {
  return this.block.zone().is_linked();
}
```

bcafuk replied:
Fixed in 506394a14afbb78b493e59d4dc5f6fc4a04890d8

MarkoZabcic commented on components/block/element.js:
```diff
@@ -9,56 +9,93 @@ export default class Block extends LitElement {
+        <div class="edit-menu">
+          <button @click=${this.refresh}>Refresh</button>
+        </div>
+      `;
```
extract to a separate function `renderMethodName`. MethodName is just placeholder

bcafuk replied:
Fixed in 506394a14afbb78b493e59d4dc5f6fc4a04890d8

MarkoZabcic commented on components/block/element.js:
```diff
@@ -9,56 +9,93 @@ export default class Block extends LitElement {
+      );
+
+      if (parent && isBlock(parent)) {
+        return html`
+          <div class="edit-menu">
+            <button @click=${parent.edit.bind(parent)}>Edit container</button>
+            <button @click=${parent.refresh.bind(parent)}>
+              Refresh container
+            </button>
+            <button @click=${this.edit}>Edit</button>
+            <button @click=${this.refresh}>Refresh</button>
+          </div>
+        `;
```
extract to a separate function `renderMethodName`. MethodName is just placeholder

bcafuk replied:
Fixed in 506394a14afbb78b493e59d4dc5f6fc4a04890d8

MarkoZabcic commented on components/block/element.js:
```diff
@@ -9,56 +9,93 @@ export default class Block extends LitElement {
+        `;
+      } else {
+        return html`
+          <div class="edit-menu">
+            <button @click=${this.edit}>Edit</button>
+            <button @click=${this.refresh}>Refresh</button>
+          </div>
```
extract to a separate function `renderMethodName`. MethodName is just placeholder

bcafuk replied:
Fixed in 506394a14afbb78b493e59d4dc5f6fc4a04890d8

MarkoZabcic commented on components/block/element.js:
```diff
@@ -9,56 +9,93 @@ export default class Block extends LitElement {
+        </div>
+      `;
+    } else {
+      const parentId = block.attributes.parent_block_id;
+      const parent = document.querySelector(
+        `ngl-block[data-ngl-block-id="${parentId}"]`
+      );
```
extract to getter

bcafuk replied:
Fixed in 506394a14afbb78b493e59d4dc5f6fc4a04890d8

MarkoZabcic commented on components/block/element.js:
```diff
@@ -9,56 +9,93 @@ export default class Block extends LitElement {
-    const nextElement = this.nextElementSibling;
-    isBlock(nextElement) && this.parentNode.insertBefore(nextElement, this)
+  edit() {
+    const block = this.getBlock();
```
Use new getter

bcafuk replied:
Fixed in 506394a14afbb78b493e59d4dc5f6fc4a04890d8

Commit:
LAYOUTS-629 Remove unused components

MarkoZabcic commented on components/block/element.js:
```diff
@@ -9,56 +9,93 @@ export default class Block extends LitElement {
+    const block = this.getBlock();
+    block.trigger('edit');
+
+    if (!this.isSaveListenerAttached) {
+      block.on('sidebar_save:success', () => {
+        this.refresh();
+      });
+      this.isSaveListenerAttached = true;
+    }
```
This should be configured in connectedCallback and in that case you don't need `isSaveListenerAttached`.

Read more here:
https://lit.dev/docs/components/lifecycle/#connectedcallback


```js
connectedCallback() {
  super.connectedCallback();
  this.block.on('sidebar_save:success', () => {
    this.refresh();
  });
}
```

:information_source:  connectedCallback is absolutely the best thing you get for free in native web components.

bcafuk replied:
Fixed in dbab07ec298e517830df04a0d10f714841c90098

MarkoZabcic commented on components/block/element.js:
```diff
@@ -9,56 +9,93 @@ export default class Block extends LitElement {
   }
 
   async refresh() {
+    this.main.classList.add('loading');
```
instead of this create new static property

```js
 static properties = {
    loading: {state: true}
```

Now you can just do:
```js
  this.loading = true;
  let html = await this.fetch();
  this.loading = false;
```

or even better:
```js
 // Move it to fetch function
 async fetch() {
   this.loading = true;
   const resp = await fetch(window.location.href);
   this.loading = false;
   return resp.text();
  }

 // and here you can just have
 let html = await this.fetch();
```
than in main render use classMap:

https://lit.dev/docs/templates/directives/#classmap

```js
render() {
    const classes = { loading: this.loading };
    
    return html`
      <main  class=${classMap(classes)} >
        ${this.renderMenu()}
        <slot></slot>
      </main>
    `;
}
```

:warning:  don't forget to import classMap directive

bcafuk replied:
Fixed in f46ee955e776e30d4608ac5cc0b071cadb0f8ac8

Commit:
LAYOUTS-629 Extract methods and introduce getters

Commit:
LAYOUTS-629 Run Prettier

Commit:
LAYOUTS-629 Fix indentation

Commit:
LAYOUTS-629 Use `connectedCallback` to reload blocks when saved

Commit:
LAYOUTS-629 Use internal reactive state to fade loading blocks

Commit:
LAYOUTS-629 Fix inconsistent quotes

Commit:
LAYOUTS-629 Rename block ID attribute

MarkoZabcic commented on components/block/spec.js:
```diff
@@ -11,16 +11,19 @@ describe('ngl-block', () => {
-      <slot></slot>
-    </main>
-    `);
+    assert.shadowDom.equal(
```
This does not exist any more:
```html
<button>Move Up</button>
<button>Move Down</button>
```

Please update and run tests so they are meaningful and passing.

bcafuk replied:
I am not quite sure how to adapt the tests, since `connectedCallback` uses the model, which is expected to be retrieved from `window.parent`.

MarkoZabcic replied:
Oh that, I will try something in the morning

MarkoZabcic replied:
I have updated and written few basic.

@bcafuk please contact me if you have any questions

MarkoZabcic commented on components/block/element.js:
```diff
@@ -9,61 +10,119 @@ export default class Block extends LitElement {
-    isBlock(prevElement) && this.parentNode.insertBefore(this, prevElement)
+  renderLinkedBlockMenu() {
+    return html`
+      <div class="edit-menu">
```
`<div class="edit-menu">` could be moved to main `render` in all occurrences.

bcafuk replied:
Moved in fe0b4ee3e33c6f4ff7dfae10e69e64f68266fe59

MarkoZabcic commented on components/block/element.js:
```diff
@@ -9,61 +10,119 @@ export default class Block extends LitElement {
+        <button @click=${this.parentModel.edit.bind(this.parentModel)}>
+          Edit container
+        </button>
+        <button @click=${this.parentModel.refresh.bind(this.parentModel)}>
```
Create dedicated functions for `@click` example `parentEdit` and `parentRefresh` like we have `edit` and `refresh`.

Lit has some optimizations for this events so avoid bind here.

bcafuk replied:
Created the functions in fe0b4ee3e33c6f4ff7dfae10e69e64f68266fe59

MarkoZabcic commented on components/block/element.js:
```diff
@@ -9,61 +10,119 @@ export default class Block extends LitElement {
+
+  renderOuterBlockMenu() {
+    return html`
+      <div class="edit-menu">
```
:information_source:  optional as I don't know what is the end vision for actions:

When you move `edit-menu` to main render... than this function can be used inside `renderInnerBlockMenu`

bcafuk replied:
Reused the function in fe0b4ee3e33c6f4ff7dfae10e69e64f68266fe59

MarkoZabcic commented on components/block/element.js:
```diff
@@ -9,61 +10,119 @@ export default class Block extends LitElement {
-        <button @click=${this.delete}>Delete</button>
-      </nav>
-    `;
+  constructor() {
+    super();
+
+    this.loading = false;
+  }
```
You can avoid this entire constructor as `this.loading` is `undefined` by default which when converted to Boolean is same as `false`. If you prefer to explicit it set it you can keep it.

bcafuk replied:
The only reason I added the constructor is because I wanted to explicitly make it `false`, rather than have it implicitly converted. I'd prefer to keep it.

MarkoZabcic commented on components/block/element.js:
```diff
@@ -9,61 +10,119 @@ export default class Block extends LitElement {
+    }));
+  }
+
+  get parentModel() {
```
This entire method is confusing and too complicated so let's simplify it.

getter is named parentModel when we are actually just returning DOM element and right above we have used model to refere to Backbone models.

Use better naming

bcafuk replied:
Wholeheartedly agree. Renamed in fe0b4ee3e33c6f4ff7dfae10e69e64f68266fe59

MarkoZabcic commented on components/block/element.js:
```diff
@@ -9,61 +10,119 @@ export default class Block extends LitElement {
+  get parentModel() {
+    if (this.cached_parent_model !== undefined) return this.cached_parent_model;
+
+    const parentId = this.model.attributes.parent_block_id;
+    const parentModel = document.querySelector(
+      `ngl-block[blockId="${parentId}"]`
+    );
```
extract this into dedicated getter

bcafuk replied:
I am not quite sure what you mean here -- A separate, non-caching getter for the parent block element?

MarkoZabcic replied:
With all new changes I've added a different suggestion.

MarkoZabcic commented on components/block/element.js:
```diff
@@ -9,61 +10,119 @@ export default class Block extends LitElement {
+  }
+
+  get parentModel() {
+    if (this.cached_parent_model !== undefined) return this.cached_parent_model;
```
`if (this.cached_parent_model !== undefined)`
change to:
`if (this.cached_parent_model)`

No need to explicitly use `undefined`

bcafuk replied:
This is to distinguish between the cases where there is no value cached (`undefined`) and when a block has no parent (`null`).

MarkoZabcic replied:
ok, can you point me to code where that is used?

bcafuk replied:
Whether the returned value is `null` is used to determine which menu to render:
https://github.com/netgen-layouts/layouts-ui-components/blob/ddfec7b6db6863d2f111e70692f5000feb38a64f/components/block/element.js#L58

MarkoZabcic replied:
So actually we only care if it's set so we can `renderInnerBlockMenu`.

So we can simplify it as `undefined` and `null` are both `false`

MarkoZabcic commented on components/block/element.js:
```diff
@@ -9,61 +10,119 @@ export default class Block extends LitElement {
+      `ngl-block[blockId="${parentId}"]`
+    );
+
+    if (parentModel && isBlock(parentModel))
```
Why do we need to test isBlock here when we are always searching for `ngl-block` which is instance of `Block`?

bcafuk replied:
You are right, I don't think it is necessary. Removed the test in fe0b4ee3e33c6f4ff7dfae10e69e64f68266fe59

MarkoZabcic replied:
Great, then you can also remove `isBlock` function on top

bcafuk replied:
Yep, I missed that. Removed `isBlock` in f4602e9bd9225f12b6fb88d62b34ba95b82b62eb

MarkoZabcic commented on components/block/element.js:
```diff
@@ -9,61 +10,119 @@ export default class Block extends LitElement {
-    const resp = await fetch(window.location.href)
-    return resp.text()
+    this.loading = true;
+    const resp = await fetch(window.location.href);
```
Add a task for the future to handle fetch failure.

MarkoZabcic replied:
Have you added task in your JIRA or somewhere to address this in some other PR?

bcafuk replied:
@MarkoZabcic Yep, I've noted it a JIRA ticket.

Commit:
LAYOUTS-629 Refactor block component code

MarkoZabcic commented on components/block/element.js:
```diff
@@ -9,62 +10,112 @@ export default class Block extends LitElement {
-    console.log(e.target.innerText)
-    const prevElement = this.previousElementSibling;
-    isBlock(prevElement) && this.parentNode.insertBefore(this, prevElement)
+  get parent() {
```
I think we can just simplfy it to:

```js
get parent() {
  const parentId = this.model.attributes.parent_block_id;
  this.cached_parent ||= document.querySelector(`ngl-block[blockId="${parentId}"]`);

  return this.cached_parent;
}
```

bcafuk replied:
While this is much more readable than my code, your suggestion however does not distinguish between the cases of "nothing cached yet" and "cached that there is no parent." The suggested function would re-run `document.querySelector` every time the getter is called on the same non-nested block.

My original implementation used `undefined` to mark an empty cache, and `null` to mark that the getter has already been called, and that no parent was found. This suggestion uses the same values*, but it does not differentiate between them to avoid unnecessarily calling `document.querySelector` again.

I definitely agree my code had readability issues, and using `undefined` and `null` for different purposes is not as clear as I would have liked. I'd appreciate some more guidance on this issue and how to resolve it.

-----
\* `document.querySelector` returns `null` for an empty match, and `undefined || null` is `null`, so the behavior is the same: `cached_parent` is `undefined` for an uninitialized cache, `null` for no parent

MarkoZabcic replied:
Thanks for explanation. In that case I suggest something like this:

```js
get parent() {
  const parentId = this.model.attributes.parent_block_id;
  this.cached_parent ||= parentId && document.querySelector(`ngl-block[blockId="${parentId}"]`);

  return this.cached_parent;
}
```

Here we weight of between code complication and optimization. But optimization for this is in microseconds or less, and even if we have it without caching concept it would still be good implementation as this part of code is super fast.

So in my last suggestion we just avoid searching for parent if it's `null` and we are ok with cases when `parent_block_id` is present but element is not.

bcafuk replied:
Yep, this looks good to me -- I have added your solution in 3924c4dc6acef5ee59e71d06904db0effaf52eed.

MarkoZabcic commented on components/block/element.js:
```diff
@@ -9,62 +10,112 @@ export default class Block extends LitElement {
+  }
+
+  renderMenu() {
+    if (this.isInLinkedZone) return this.renderLinkedBlockMenu();
+    else if (this.parent) return this.renderInnerBlockMenu();
+    else return this.renderOuterBlockMenu();
```
Because you are using `return` you can also simplify this to:

```js
 if (this.isInLinkedZone) return this.renderLinkedBlockMenu();
 if (this.parent) return this.renderInnerBlockMenu();

 return this.renderOuterBlockMenu();
```

There is no point of `else return`

bcafuk replied:
Removed in f4602e9bd9225f12b6fb88d62b34ba95b82b62eb

MarkoZabcic commented on components/block/element.js:
```diff
@@ -9,62 +10,112 @@ export default class Block extends LitElement {
+    );
+  }
+
+  async parentRefresh() {
+    await this.parent.refresh();
```
Remove async, await as it's not needed here.

await is only needed if you do something with promise response, and async is only needed if you use await inside the function.

bcafuk replied:
Removed in f4602e9bd9225f12b6fb88d62b34ba95b82b62eb.
The promise from `parent.refresh()` can can easily be returned from `parentRefresh` if it ever becomes necessary to `await` it at some point.

MarkoZabcic replied:
Exactly!

MarkoZabcic committed:
feat: Update tests and write base cases

- Create helper function to stub layout environment
- Introduce sinon so we can do advance stubs/mocks/spies

Commit:
LAYOUTS-629 Refactor block component code some more

Commit:
LAYOUTS-629 Simplify parent getter

MarkoZabcic reviewed (approved):
This looks good now :+1:

emodric commented:
Thanks @bcafuk @MarkoZabcic , excellent work!
